src/raw/mod.rs

```diff
@@ -0,0 +1,1634 @@
+    ///   * the last byte of `data` is a 0
+    ///
+    /// Note that the internal structure of the bytes representing the
+    /// BSON elements is _not_ validated at all by this method. If the
+    /// bytes do not conform to the BSON spec, then method calls on
+    /// the RawDocument will return RawErrors where appropriate.
+    ///
+    /// ```
+    /// # use bson::raw::{RawDocument, RawError};
+    /// let doc = RawDocument::new(b"\x05\0\0\0\0".to_vec())?;
+    /// # Ok::<(), RawError>(())
+    /// ```
+    pub fn new(data: Vec<u8>) -> RawResult<RawDocument> {
+        if data.len() < 5 {
+            return Err(RawError::MalformedValue("document too short".into()));
+        }
+
+        let length = i32_from_slice(&data[..4]);
+
+        if data.len() as i32 != length {
+            return Err(RawError::MalformedValue("document length incorrect".into()));
+        }
+
+        if data[data.len() - 1] != 0 {
+            return Err(RawError::MalformedValue(
+                "document not null-terminated".into(),
+            ));
+        }
+
+        Ok(Self {
+            data: data.into_boxed_slice(),
+        })
+    }
+
+    /// Create a RawDocument from a Document.
+    ///
+    /// ```
+    /// # use bson::raw::{RawDocument, RawError};
+    /// use bson::{doc, oid::ObjectId};
+    ///
+    /// let document = doc! {
+    ///     "_id": ObjectId::new(),
+    ///     "name": "Herman Melville",
+    ///     "title": "Moby-Dick",
+    /// };
+    /// let doc = RawDocument::from_document(&document);
+    /// # Ok::<(), RawError>(())
+    /// ```
+    pub fn from_document(doc: &Document) -> RawDocument {
+        let mut data = Vec::new();
+        doc.to_writer(&mut data).unwrap();
+
+        Self {
+            data: data.into_boxed_slice(),
+        }
+    }
+
+    /// Gets an iterator over the elements in the `RawDocument`, which yields `Result<&str,
+    /// Element<'_>>`.
+    ///
+    /// ```
+    /// # use bson::raw::{elem, RawDocument, RawError};
+    /// use bson::doc;
+    ///
+    /// let doc = RawDocument::from_document(&doc! { "ferris": true });
+    ///
+    /// for element in doc.iter() {
+    ///     let (key, value) = element?;
+    ///     assert_eq!(key, "ferris");
+    ///     assert_eq!(value.as_bool()?, true);
+    /// }
+    /// # Ok::<(), RawError>(())
+    /// ```
+    ///
+    /// # Note:
+    ///
+    /// There is no owning iterator for RawDocument.  If you need ownership over
+    /// elements that might need to allocate, you must explicitly convert
+    /// them to owned types yourself.
+    pub fn iter(&self) -> RawDocumentIter<'_> {
+        self.into_iter()
+    }
+
+    /// Return the contained data as a `Vec<u8>`
+    ///
+    /// ```
+    /// # use bson::raw::RawDocument;
+    /// use bson::doc;
+    ///
+    /// let doc = RawDocument::from_document(&doc!{});
+    /// assert_eq!(doc.into_inner(), b"\x05\x00\x00\x00\x00".to_vec());
+    /// ```
+    pub fn into_inner(self) -> Vec<u8> {
```

Comment:
.into_inner is a naming convention I [borrowed from the standard library](https://doc.rust-lang.org/std/?search=into_inner), but into_vec() is also common.  I do think it should exist as an inherent method by one of those names, since it's fundamental to the data structure itself that it is a wrapper around heap allocated bytes, and not just some arbitrary conversion.